FFmpegInterop/FLACSampleProvider.cpp

```diff
@@ -28,24 +30,41 @@ namespace winrt::FFmpegInterop::implementation
 	FLACSampleProvider::FLACSampleProvider(_In_ const AVFormatContext* formatContext, _In_ AVStream* stream, _In_ Reader& reader) :
 		SampleProvider(formatContext, stream, reader)
 	{
+		THROW_HR_IF(MF_E_INVALID_FILE_FORMAT, m_stream->codecpar->extradata_size < FLAC_STREAMINFO_SIZE);
 
+		// Check if the codec private data includes the FLAC marker and block header before the FLAC stream info
+		if (equal(FLAC_MARKER.begin(), FLAC_MARKER.end(), m_stream->codecpar->extradata))
+		{
+			THROW_HR_IF(MF_E_INVALID_FILE_FORMAT, m_stream->codecpar->extradata_size < FLAC_MARKER.size() + FLAC_STREAMINFO_HEADER.size() + FLAC_STREAMINFO_SIZE);
+			m_flacStreamInfo = m_stream->codecpar->extradata + FLAC_MARKER.size() + FLAC_STREAMINFO_HEADER.size();
+		}
+		else
+		{
+			m_flacStreamInfo = m_stream->codecpar->extradata;
+		}
+
+		// Update the codec parameters based on the stream info. These values may not have been set if FFmpeg wasn't built with the FLAC decoder.
+		BitstreamReader bitstreamReader{ m_flacStreamInfo, FLAC_STREAMINFO_SIZE };
+		bitstreamReader.SkipN(16); // Min block size
+		bitstreamReader.SkipN(16); // Max block size
+		bitstreamReader.SkipN(24); // Min frame size
+		bitstreamReader.SkipN(24); // Max frame size
+		m_stream->codecpar->sample_rate = static_cast<int>(bitstreamReader.ReadN(20));
+		m_stream->codecpar->channels = static_cast<int>(bitstreamReader.ReadN(3)) + 1;
+		m_stream->codecpar->bits_per_raw_sample = static_cast<int>(bitstreamReader.ReadN(5)) + 1;
+		// Remaining fields left unparsed as they are unneeded at this time
 	}
 
 	tuple<IBuffer, int64_t, int64_t, vector<pair<GUID, IInspectable>>, vector<pair<GUID, IInspectable>>> FLACSampleProvider::GetSampleData()
 	{
 		if (m_isDiscontinuous)

```

Comment:
Is this using the MS decoder? Unless the decoder was unloaded it will have an audio glitch if you resend the data.

Reply:
Yes, it's the MS decoder. This is the same behavior as our in-box MKV source and I don't notice any glitches when seeking during playback.

Reply:
We implemented that behavior in the FLAC source and it broke it. So we did the Out of Band media type to pass the information and it fixed that issue.

---
In reply to: [419703605](https://github.com/microsoft/FFmpegInterop/pull/274#discussion_r419703605) [](ancestors = 419703605)

Reply:
However if you are not seeing an issue it might be because you are just sending the stream info and not the entire header. Though if that is the case what happens in Groove when you insert the equalizer effect?

---
In reply to: [419706180](https://github.com/microsoft/FFmpegInterop/pull/274#discussion_r419706180) [](ancestors = 419706180,419703605)

Reply:
The media type attribute used by the FLAC source is private. Not sure why we did that instead of using MF_MT_USER_DATA...

I'm not seeing any issues, but if something comes up we'll need a publicly available solution.